### PR TITLE
Deprecate config.Unmarshallable in favor of confmap.Unmarshallable

### PR DESCRIPTION
**Pull request by djaglowski**
This interface is concerned with unmarshalling a `confmap.Conf`, so is arguably more appropriate
in the `confmap` package. More concretely, moving this is necessary to avoid an import cycle as
would be introduced by #6029.

### CODE REVIEW

File: confmap/confmap.go [hunk 1]
```diff
@@ -209,3 +209,11 @@ func mapKeyStringToMapKeyTextUnmarshalerHookFunc() mapstructure.DecodeHookFuncTy
 		return data, nil
 	}
 }
+
+// Unmarshallable defines an optional interface for custom configuration unmarshalling.
+// A configuration struct can implement this interface to override the default unmarshalling.
+type Unmarshallable interface {
```
Comment (bogdandrutu) [inline confmap/confmap.go:215]: s/Unmarshallable/Unmarshaler I think this is the standard that almost all the golang is using. I know the name is my fault, but good opportunity to fix it.

File: confmap/confmap.go [hunk 2]
```diff
@@ -209,3 +209,11 @@ func mapKeyStringToMapKeyTextUnmarshalerHookFunc() mapstructure.DecodeHookFuncTy
 		return data, nil
 	}
 }
+
+// Unmarshallable defines an optional interface for custom configuration unmarshalling.
+// A configuration struct can implement this interface to override the default unmarshalling.
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply confmap/confmap.go:214]: I don't have a good sense of the technical difference between the terms. Is there a standard way to decide?

[the remark this answers is not in the document]
Comment (bogdandrutu) [reply confmap/confmap.go]: Not that I know, because if I've known I would have applied that :)) So happy to hear ideas/opinions here.

File: CHANGELOG.md [hunk 3]
```diff
@@ -26,11 +26,8 @@
 - Deprecate `pcommon.New[Trace|Span]ID` in favor direct conversion. (#6008)
 - Deprecate `MetricDataPointFlagsImmutable` type. (#6017)
 - Deprecate `*DataPoint.[Set]FlagsImmutable()` funcs in favor of `*DataPoint.[Set]Flags()`. (#6017)
-
-### 🚩 Deprecations 🚩
-
 - Deprecate `LogRecord.FlagsStruct()` and `LogRecord.SetFlagsStruct()` in favor of `LogRecord.Flags()` and `LogRecord.SetFlags()`. (#6007)
-
+- Deprecate `config.Unmarshallable` in favor of `confmap.Unmarshaler`. (#6031)
 ### 💡 Enhancements 💡
```
Comment (bogdandrutu) [inline CHANGELOG.md:31]: Need an empty line, also rebase.

File: CHANGELOG.md [hunk 4]
```diff
@@ -69,7 +66,7 @@
   - `NewMetricDataPointFlagsStruct` -> `NewMetricDataPointFlags`
 - Deprecate builder distribution flags, use configuration. (#5946)
 - Enforce naming conventions for Invalid[Trace|Span]ID: (#5969)
-  - Deprecate funcs `pcommon.InvalidTraceID` and `pcommon.InvalidSpanID` in favor of vars `pcommon.EmptyTraceID` and `pcommon.EmptySpanID` 
+  - Deprecate funcs `pcommon.InvalidTraceID` and `pcommon.InvalidSpanID` in favor of vars `pcommon.EmptyTraceID` and `pcommon.EmptySpanID`
```
Comment (bogdandrutu) [inline CHANGELOG.md:69]: Good formatting changes, please send me a separate PR with these.

File: CHANGELOG.md [hunk 5]
```diff
@@ -212,7 +209,7 @@ There isn't a valid core binary for this release. Use v0.57.2 instead.
   - `component.WithTracesProcessorAndStabilityLevel` -> `component.WithTracesProcessor`
   - `component.WithMetricsProcessorAndStabilityLevel` -> `component.WithMetricsProcessor`
   - `component.WithLogsProcessorAndStabilityLevel` -> `component.WithLogsProcessor`
-  - 
+  -
```
Comment (bogdandrutu) [inline CHANGELOG.md:212]: The `-` needs to be removed.

File: confmap/confmap.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (djaglowski) [reply confmap/confmap.go]: To me, in the context of transforming a generic data format to a specific struct, they are basically interchangeable terms. That said, I think implementing this interface is quite similar to the pattern of implementing json or yaml unmarshaling interfaces, so I would suggest we stick with "unmarshal".

File: CHANGELOG.md (hunk 4 above)
Comment (djaglowski) [reply CHANGELOG.md:69]: Oops, autoformatting from my IDE. Anyways, I moved them here: https://github.com/open-telemetry/opentelemetry-collector/pull/6035

File: confmap/confmap.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply confmap/confmap.go]: In the current world we have 3 parts:

1. The config provider. This the "Format and Location" of the config (yaml/json).
2. The in-memory raw configuration. This is the `confmap.Conf`. The process of "looking for word" to transform from raw configuration (`map[string]interface{}`) into the Config object is the process we are looking for.
3. The final Config.

So loading a config is a 2 steps: 

`Location/Format -> Confmap -> Config`

I tried to explain also to others what we have, and we are looking for a word/terminology for `Confmap -> Config`. Probably Unmarshal/Marshal is not bad but want to make sure we explain to others, so may get more ideas.

**Comment by bogdandrutu** [comment]
Please rebase

File: confmap/confmap.go [hunk 6]
```diff
@@ -209,3 +209,10 @@ func mapKeyStringToMapKeyTextUnmarshalerHookFunc() mapstructure.DecodeHookFuncTy
 		return data, nil
 	}
 }
+
+// Unmarshaler interface may be implemented by types to customize their behavior when being unmarshaled from a Conf.
+type Unmarshaler interface {
+	// Unmarshal is a function that unmarshals a confmap.Conf into the struct in a custom way.
+	// The confmap.Conf for this specific component may be nil or empty if no config available.
```
Comment (mx-psi) [inline confmap/confmap.go:216]: `s/confmap.Conf/Conf` and some minor rewording
```suggestion
	// Unmarshal a Conf into the struct in a custom way.
	// The Conf for this specific component may be nil or empty if no config available.
```

**Review by bogdandrutu — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

File: confmap/confmap.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (djaglowski) [reply confmap/confmap.go]: I brought this up at the SIG to get other opinions. I suggest we give this a couple more hours and then merge if there are no objections.

[the remark this answers is not in the document]
Comment (djaglowski) [reply confmap/confmap.go]: There do not appear to be alternate opinions so I believe this is ready to merge.
